chore: create appium monorepo; closes #14441

boneskull opened the pull request:
## UPDATE - See https://github.com/appium/appium/pull/15343#issuecomment-843529470

_This PR supersedes PR #14562, but ought to be very similar._

## Proposed changes

This is my first attempt at making this repo a monorepo and moving appium-proper into `packages/appium`.

I had tried to make this work with npm's "workspaces", but it seems lifecycle scripts don't run as you'd expect them to upon install.  Instead, I'm using Lerna, which provides more features for monorepos.

Changes include:

- The root `package.json` is now a package named `appium-monorepo`, which is `private`.  The monorepo should not be published.
- All dev dependencies remain in the root `package.json`.
- `gulpfile.js` _moved_ into `packages/appium/gulpfile.js`, but this should be **temporary**--`appium-gulp-plugins` makes certain assumptions which a monorepo violates, and I was unable to get it working without patching `appium-gulp-plugins`.
- Most of the `package.json` `scripts` moved into `packages/appium/package.json` due to the above.  This too should be considered **temporary**.
- Adds a root `postinstall` lifecycle script which calls `lerna bootstrap` to install/link packages.  I don't see a `package-lock.json` in `packages/appium/`; this needs investigating.
- Adds a root `test` lifecycle script which executes `npm run test` in all packages.  The tests do not pass, but I don't know if this is my fault.
- The `clean` script will purge all `node_modules` folders in each package, as well.
- I removed `scripts/release*.sh`, since it seems they would fail anyway--`lerna` should be used for releases instead
- The shrinkwrap-related stuff is untested.
- Renamed `tags` prop in `packages/appium/package.json` to `keywords`.  I don't know if `tags` is supported.
- Removed an unused `projectRoot` prop in the gulpfile.

## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/appium/appium/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla.js.foundation/appium/appium)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

jlipps committed:
Merge pull request #314 from appium/jlipps/misc

add isArray to cap validation strategies

jlipps committed:
fix flakey unexpected shutdown test

jlipps committed:
bump version to 3.15.5

jlipps committed:
Merge pull request #315 from appium/jlipps/fixflake

fix flakey unexpected shutdown test

jlipps committed:
factor out image element registration to helper function (#312)

...so the function can be used by other drivers / plugins

mykola-mokhnach committed:
Use proper Error object when throwing download failures (#316)

dpgraham committed:
fix: Make proxy convert /property to /attribute (#311)

imurchie committed:
Be more explicit about unrecognized caps (#317)

greenkeeper[bot] committed:
chore(package): update chromedriver to version 73.0.0 (#318)

greenkeeper[bot] committed:
chore(package): update chromedriver to version 74.0.0 (#319)

imurchie committed:
Remove chromedriver dev dep (#320)

KazuCocoa committed:
fix handling 0 in gsmSignal (#321)

* fix handling 0 in gsmSignal

* use hasValue

greenkeeper[bot] committed:
fix(package): update ws to version 7.0.0 (#322)

greenkeeper[bot] committed:
Update appium-gulp-plugins to the latest version 🚀 (#323)

* chore(package): update appium-gulp-plugins to version 4.0.0

* Move to test on Node 10 and 12

imurchie committed:
Extend unexpected timeout test delay

imurchie committed:
Only get message from error (#324)

Dan-Maor committed:
added keepalive option to JWProxy (#325)

imurchie committed:
3.16.0

KazuCocoa committed:
add getting result image by find by image (#327)

* add getMatchedImageResult and return the result if it is true

* add tests

* tweak error case

greenkeeper[bot] committed:
fix(package): update validate.js to version 0.13.0 (#332)

Siddh010 committed:
Better message for creating session with W3C (#331)

jlipps committed:
Add endpoint that receives a webdriver script for batch command execution (#328)

* add executeDriverScript route, command, tests

* allow driver scripts to write and collect console logs

* ensure driver script protocol is set to same protocol as driver, and also includes all appium commands

* implement isFeatureEnabled to check for insecure feature

* add tests for isFeatureEnabled

* update feature flag name to match style of other flags

* add ensureFeatureEnabled method for use by other drivers

* move driver script execution into a child process so we can kill it if necessary

* use this.opts.address instead of this.opts.host

* ensure driver script timeout is a number in the child proc as well

* ensure we get webdriverio without issues

jlipps committed:
3.17.0

mykola-mokhnach committed:
Optimize memory usage by proxy class (#334)

mykola-mokhnach committed:
3.17.1

vrunoa committed:
feat: Add parseCapsArray to base driver (#333)

Commit:
Release 3.18.0

mykola-mokhnach committed:
Add a workaround to POST /window W3C call (#335)

mykola-mokhnach committed:
Update the logic that determines the session protocol (#336)

mykola-mokhnach committed:
3.19.0

mykola-mokhnach committed:
Improve the logic of proxyReqRes method (#337)

imurchie committed:
Fix travis build (#339)

mykola-mokhnach committed:
Add protocol conversion for setFrame call (#340)

mykola-mokhnach committed:
3.19.1

imurchie committed:
Allow CORS for receiving async responses (#342)

imurchie committed:
3.20.0

sa-spag committed:
Handle Content-Type with parameter (#341)

mykola-mokhnach committed:
Update response handling for proxyReqRes call (#344)

mykola-mokhnach committed:
3.20.1

KazuCocoa committed:
accept text for setValueImmediate and replaceValue as W3C (#347)

mykola-mokhnach committed:
fix: Properly log request errors (#348)

mykola-mokhnach committed:
3.20.2

mykola-mokhnach committed:
chore: Export isStandardCap method (#349)

mykola-mokhnach committed:
3.21.0

mykola-mokhnach committed:
fix: Fix session identifier retrieval for W3C protocol (#351)

mykola-mokhnach committed:
3.21.1

mykola-mokhnach committed:
fix: Do not print stack for request errors (#352)

mykola-mokhnach committed:
3.21.2

imurchie committed:
refactor: make server parameters a hash [breaking change] (#350)

jlipps committed:
feat: stop hardcoding /wd/hub and make route prefix configurable [BREAKING CHANGE] (#346)

jlipps committed:
4.0.0

mykola-mokhnach committed:
fix: Patch downstream responses for unknown upstream protocol (#353)

mykola-mokhnach committed:
4.0.1

imurchie committed:
fix: remove redundant log line on cached app found (#354)

mykola-mokhnach committed:
fix: Tune status presence verification for generic protocol (#355)

mykola-mokhnach committed:
4.0.2

imurchie committed:
fix: adjust position of send button on guinea-pig pages

imurchie committed:
4.0.3

mykola-mokhnach committed:
fix: Avoid using instanceof verification (#356)

imurchie committed:
fix: fix the build for Node 12.11.0 (#359)

mykola-mokhnach committed:
fix: Adjust the error codes for NotImplementedError (#357)

mykola-mokhnach committed:
4.1.0

mykola-mokhnach committed:
refactor: Replace custom commands queue implementation with async lock (#362)

mykola-mokhnach committed:
4.2.0

mykola-mokhnach committed:
refactor: Reuse walkDir implementation from appium-support module (#361)

mykola-mokhnach committed:
fix: Make device name non required (#363)

mykola-mokhnach committed:
4.2.1

jlipps committed:
feat: add new route to allow logging of custom events (#364)

jlipps committed:
4.3.0

jlipps committed:
feat: add ability to retrieve log events as a new api route (#365)

jlipps committed:
4.4.0

mykola-mokhnach committed:
fix: Explicitly set keep-alive server timeout (#366)

mykola-mokhnach committed:
fix: Fix the request options for different keepAlive values (#367)

mykola-mokhnach committed:
4.4.1

KazuCocoa committed:
feat: add filtering log events with type (related to #365) (#368)

* add filtering with type

* use reduce, add converting to array

* make anonymous

KazuCocoa committed:
4.5.0

mykola-mokhnach committed:
fix: Sync parameters validation for the old and the new set alert text endpoints (#369)

mykola-mokhnach committed:
4.5.1

mykola-mokhnach committed:
refactor: Prefer event emitting to promises cancelation (#370)

BREAKING CHANGE: The current refactoring is breaking and changes `onUnexpectedShutdown` property to a function, which accepts a handler that is invoked by the driver when it gets shut down unexpectedly

mykola-mokhnach committed:
5.0.0

mykola-mokhnach committed:
fix: make sure the listener gets removed after the command executor is completed (#371)

mykola-mokhnach committed:
5.0.1

imurchie committed:
fix: when parsing cap array, pass array through (#373)

imurchie committed:
5.0.2

mykola-mokhnach committed:
refactor: Change the setting update validation, so we don't print unnecessary warnings (#372)

mykola-mokhnach committed:
refactor: Change the applications caching logic (#374)

mykola-mokhnach committed:
5.1.0

mykola-mokhnach committed:
fix: Perform synchronous removal of applications cache (#375)

mykola-mokhnach committed:
5.1.1

mykola-mokhnach committed:
chore: Import rimrafSync from appium-support (#376)

imurchie committed:
refactor: move to new timing API (#377)

imurchie committed:
5.1.2

greenkeeper[bot] committed:
chore(package): update sinon to version 8.0.0 (#378)

greenkeeper[bot] committed:
chore(package): update appium-gulp-plugins to version 5.0.0 (#379)

mykola-mokhnach committed:
chore: Tune logging for applications cache items disposal (#380)

imurchie committed:
refactor: just a little more visibility into unprefixed caps (#381)

mykola-mokhnach committed:
5.1.3

imurchie committed:
fix: terminate socket on DELETE and NoSuchDriverError (#382)

imurchie committed:
5.2.0

imurchie committed:
fix: add another condition for terminating socket

imurchie committed:
5.2.1

greenkeeper[bot] committed:
Update mocha to the latest version 🚀 (#385)

* chore(package): update mocha to version 7.0.1

* try: change chrome config on travis

Co-authored-by: Isaac A. Murchie <isaac.appium@gmail.com>

imurchie committed:
refactor: use util.pluralize (#386)

KazuCocoa committed:
fix: remove test/ build/test for notarization to fix a warning message (#387)

* remove test/ build/test for notarization

* remove ,

* ignore only fixtures

KazuCocoa committed:
5.3.0

greenkeeper[bot] committed:
chore(package): update sinon to version 9.0.0 (#388)

mykola-mokhnach committed:
chore: Simplify proxy logging flow (#389)

mykola-mokhnach committed:
5.3.1

imurchie committed:
feat: add a simple static page with shadow elements (#390)

imurchie committed:
fix: add style to shadow element, to trigger circular structure

imurchie committed:
5.4.0

mykola-mokhnach committed:
chore: Import uuid from appium-suuport (#391)

mykola-mokhnach committed:
5.4.1

mykola-mokhnach committed:
chore: Add width/height verification to avoid OOM on image resize (#393)

mykola-mokhnach committed:
refactor: Simplify the logic of unzipApp helper (#392)

mykola-mokhnach committed:
5.4.2

jsa34 committed:
fix: Add button to actions payload (#395)

mykola-mokhnach committed:
chore: Use file name sanitizer imported from appium-support (#394)

mykola-mokhnach committed:
5.4.3

KazuCocoa committed:
feat: add score attribute for image element (#396)

* feat: add score

* add doc

* bump support

* fix docstring

KazuCocoa committed:
5.5.0

mykola-mokhnach committed:
feat: Consider the content of Cache-Control header for application caching (#397)

mykola-mokhnach committed:
5.6.0

dependabot-preview[bot] committed:
chore(deps): bump webdriverio from 5.22.4 to 6.0.2 (#399)

Bumps [webdriverio](https://github.com/webdriverio/webdriverio) from 5.22.4 to 6.0.2.
- [Release notes](https://github.com/webdriverio/webdriverio/releases)
- [Changelog](https://github.com/webdriverio/webdriverio/blob/v6.0.2/CHANGELOG.md)
- [Commits](https://github.com/webdriverio/webdriverio/compare/v5.22.4...v6.0.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

mykola-mokhnach committed:
feat: Add requests idempotency verification on the server side (#400)

mykola-mokhnach committed:
5.7.0

imurchie committed:
fix: put full date on the static guinea pig pages (#402)

imurchie committed:
chore: fix linting issues (#403)

* chore: fix linting issues

* fix: address review

mykola-mokhnach committed:
fix: Avoid ERR_STREAM_WRITE_AFTER_END error (#401)

mykola-mokhnach committed:
5.7.1

imurchie committed:
chore: move to dependabot (#404)

KazuCocoa committed:
feat: add calling chrome devtools endpoint for chromium (#405)

* add calling chrome devtools for chromium

* fill command and required parameter in

* replace goog with :vendor

KazuCocoa committed:
5.8.0

snoopy-peace committed:
chore: Define options for for doubleClick, buttonUp\Down routes (#406)

mykola-mokhnach committed:
chore: Switch to axios in App download (#407)

mykola-mokhnach committed:
5.8.1

mykola-mokhnach committed:
chore: Replace request usage with axios in tests (#408)

mykola-mokhnach committed:
refactor: Use axios for the actual proxy implementation instead of the deprecated request module (#409)

BREAKING CHANGE: The request method of Proxy class now accepts and returns axios-specific values
BREAKING CHANGE: The default value of keepAlive setting has been changed to true

mykola-mokhnach committed:
6.0.0

mykola-mokhnach committed:
fix: Adjust the validation of proxied requests (#410)

mykola-mokhnach committed:
6.0.1

mykola-mokhnach committed:
refactor: Move common constants to a separate file to avoid unnecessary cross-imports (#411)

mykola-mokhnach committed:
chore: do not print full stack traces for connection errors (#412)

mykola-mokhnach committed:
6.1.0

DudaGod committed:
fix(welcome page): add "viewport" meta tag (#413)

- add "viewport" meta tag in order to correctly determine viewport width and height in safari

mykola-mokhnach committed:
chore: Tune the target folder name for an extracted app

mykola-mokhnach committed:
fix: Make sure getLogTypes returns a promise (#414)

mykola-mokhnach committed:
6.2.0

mykola-mokhnach committed:
chore: Print a compact message for 'socket hang up' error (#415)

mykola-mokhnach committed:
6.2.1

mykola-mokhnach committed:
fix: Properly assign connection pools for JW proxy (#416)

mykola-mokhnach committed:
6.2.2

dependabot-preview[bot] committed:
chore(deps-dev): bump mocha from 7.2.0 to 8.0.1 (#417)

Bumps [mocha](https://github.com/mochajs/mocha) from 7.2.0 to 8.0.1.
- [Release notes](https://github.com/mochajs/mocha/releases)
- [Changelog](https://github.com/mochajs/mocha/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mochajs/mocha/compare/v7.2.0...v8.0.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

mykola-mokhnach committed:
fix: Enforce UTF-8 encoding on extracted file names for .ipa archives (#419)

mykola-mokhnach committed:
6.2.3

dependabot-preview[bot] committed:
chore(deps): bump lru-cache from 5.1.1 to 6.0.0 (#420)

Bumps [lru-cache](https://github.com/isaacs/node-lru-cache) from 5.1.1 to 6.0.0.
- [Release notes](https://github.com/isaacs/node-lru-cache/releases)
- [Commits](https://github.com/isaacs/node-lru-cache/compare/v5.1.1...v6.0.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

mykola-mokhnach committed:
chore: Never apply automatic proxy config to JWProxy requests (#421)

jlipps committed:
feat: allow plugins to modify the server and method map (#422)

jlipps committed:
7.0.0

JesseCodeBones committed:
fix: send empty string instead of null in setValue call (#423)

mykola-mokhnach committed:
7.0.1

mykola-mokhnach committed:
chore: Always destroy active sockets on server shutdown (#424)

mykola-mokhnach committed:
7.1.0

dpgraham committed:
fix: allow prefixed standard capabilities (#425)

* fix: allow prefixed standard capabilities

* PR Fix

* Add test for duplicates

dpgraham committed:
Release 7.2.0

dependabot-preview[bot] committed:
chore(deps): bump http-status-codes from 1.4.0 to 2.1.1 (#428)

* chore(deps): bump http-status-codes from 1.4.0 to 2.1.1

Bumps [http-status-codes](https://github.com/prettymuchbryce/node-http-status) from 1.4.0 to 2.1.1.
- [Release notes](https://github.com/prettymuchbryce/node-http-status/releases)
- [Commits](https://github.com/prettymuchbryce/node-http-status/compare/v1.4.0...2.1.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

* Update imports

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>
Co-authored-by: Mykola Mokhnach <mokhnach@gmail.com>

dependabot-preview[bot] committed:
chore(deps): bump axios from 0.19.2 to 0.20.0 (#426)

Bumps [axios](https://github.com/axios/axios) from 0.19.2 to 0.20.0.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/master/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v0.19.2...v0.20.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

KazuCocoa committed:
7.2.1

mykola-mokhnach committed:
fix: Disable request cancellation (#430)

mykola-mokhnach committed:
7.2.2

dependabot-preview[bot] committed:
chore: Update Dependabot config file (#431)

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

mykola-mokhnach committed:
fix: Return proper responses to unexpected errors (#432)

mykola-mokhnach committed:
7.2.3

mykola-mokhnach committed:
chore: Remove the unused ajv dependency

mykola-mokhnach committed:
chore: Remove Coveralls integration

mykola-mokhnach committed:
feat: Add Webauthn routes (#433)

mykola-mokhnach committed:
7.3.0

jlipps committed:
fix: catchall handler should be added after plugin application (#434)

jlipps committed:
7.3.1

KazuCocoa committed:
feat: Remove wd hub prefix (#429)

* remove /wd/hub

* tweak comment

mykola-mokhnach committed:
fix: Use correct property to figure out LRU cache size (#435)

mykola-mokhnach committed:
7.3.2

dependabot[bot] committed:
chore: Bump axios from 0.20.0 to 0.21.0 (#436)

Bumps [axios](https://github.com/axios/axios) from 0.20.0 to 0.21.0.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/master/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v0.20.0...v0.21.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mykola-mokhnach committed:
fix: Never destroy sockets explicitly (#437)

mykola-mokhnach committed:
7.3.3

KazuCocoa committed:
ci: Set up CI with Azure Pipelines (#439)

* ci: Set up CI with Azure Pipelines

[skip ci]

* drop travis

* tweak

* add EAI_AGAIN?

* try to remove xvfb

* remove trigger restriction

KazuCocoa committed:
chore: merge master to 2.0 branch (#438)

* fix: Use correct property to figure out LRU cache size (#435)

* 7.3.2

* chore: Bump axios from 0.20.0 to 0.21.0 (#436)

Bumps [axios](https://github.com/axios/axios) from 0.20.0 to 0.21.0.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/master/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v0.20.0...v0.21.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* fix: Never destroy sockets explicitly (#437)

* 7.3.3

Co-authored-by: Mykola Mokhnach <mokhnach@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

KazuCocoa committed:
chore: merge master into 2.0 (#441)

* fix: Use correct property to figure out LRU cache size (#435)

* 7.3.2

* chore: Bump axios from 0.20.0 to 0.21.0 (#436)

Bumps [axios](https://github.com/axios/axios) from 0.20.0 to 0.21.0.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/master/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v0.20.0...v0.21.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* fix: Never destroy sockets explicitly (#437)

* 7.3.3

* ci: Set up CI with Azure Pipelines (#439)

* ci: Set up CI with Azure Pipelines

[skip ci]

* drop travis

* tweak

* add EAI_AGAIN?

* try to remove xvfb

* remove trigger restriction

Co-authored-by: Mykola Mokhnach <mokhnach@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

KazuCocoa committed:
chore: bump major version

jlipps committed:
chore!: remove image element code since it is now handled by @appium/images-plugin

jlipps committed:
feat: make server connection timeout configurable

jlipps committed:
7.4.0

KazuCocoa committed:
feat: make vendor prefix mandatory (#440)

* fix: Use correct property to figure out LRU cache size (#435)

* 7.3.2

* chore: Bump axios from 0.20.0 to 0.21.0 (#436)

Bumps [axios](https://github.com/axios/axios) from 0.20.0 to 0.21.0.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/master/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v0.20.0...v0.21.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* fix: Never destroy sockets explicitly (#437)

* 7.3.3

* make vendor prefix mondatory

* accept only w3c spec

* tweak tests

* tweak protocol e2e test

* update e2e for w3c base

* tweak

* fix todo

* update wording, add todo

Co-authored-by: Mykola Mokhnach <mokhnach@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mykola-mokhnach committed:
fix: Avoid unhandled error on driver exit (#447)

KazuCocoa committed:
7.4.1

KazuCocoa committed:
fix: Avoid unhandled error on driver exit (#447) (#448)

Co-authored-by: Mykola Mokhnach <mokhnach@gmail.com>

jlipps committed:
Merge branch 'master' into 2.0

jlipps committed:
chore: bump to 8.0.0-beta.2 since beta.1 is already published

jlipps committed:
feat: allow retrieval of basedriver version from any inheriting driver

jlipps committed:
chore: bump to 8.0.0-beta.3

jlipps committed:
fix: update plugin server adjustment to use new style

jlipps committed:
chore: bump to 8.0.0-beta.4

jlipps committed:
fix!: remove image element proxy check, and allow proxyRouteIsAvoided to inspect the body

jlipps committed:
chore: bump 2.0 to beta.5

jlipps committed:
refactor!: generalize concepts of server updater and extra method map
basedriver shouldn't really know about plugins or drivers.
it should just expose the ability to update the server or add methods

jlipps committed:
chore: bump 8.0 beta to 8.0.0-beta.6

boneskull committed:
chore: create appium monorepo; closes #14441

This PR supersedes PR #14562.

This is my first attempt at making this repo a monorepo and moving appium-proper into `packages/appium`.

I had tried to make this work with npm's "workspaces", but it seems lifecycle scripts don't run as you'd expect them to upon install.  Instead, I'm using Lerna, which provides more features for monorepos.

Changes include:

- The root `package.json` is now a package named `appium-monorepo`, which is `private`.  The monorepo should not be published.
- All dev dependencies remain in the root `package.json`.
- `gulpfile.js` _moved_ into `packages/appium/gulpfile.js`, but this should be **temporary**--`appium-gulp-plugins` makes certain assumptions which a monorepo violates, and I was unable to get it working without patching `appium-gulp-plugins`.
- Most of the `package.json` `scripts` moved into `packages/appium/package.json` due to the above.  This too should be considered **temporary**.
- Adds a root `postinstall` lifecycle script which calls `lerna bootstrap` to install/link packages.  I don't see a `package-lock.json` in `packages/appium/`; this needs investigating.
- Adds a root `test` lifecycle script which executes `npm run test` in all packages.  The tests do not pass, but I don't know if this is my fault.
- The `clean` script will purge all `node_modules` folders in each package, as well.
- I removed `scripts/release*.sh`, since it seems they would fail anyway--`lerna` should be used for releases instead
- The shrinkwrap-related stuff is untested.
- Renamed `tags` prop in `packages/appium/package.json` to `keywords`.  I don't know if `tags` is supported.

boneskull commented:
Looks like I need to sort out the `package-lock.json` issue w/ Lerna.

boneskull commented:
Next steps:

1. [ ] Fix CI / Lerna
2. [ ] Move `appium-gulp-plugins` into the monorepo.
3. [ ] Fix `appium-gulp-plugins` to work properly with this setup
4. [ ] Move scripts and `gulpfile.js` back into root.
5. [ ] Get tests passing

boneskull committed:
update node.js workflow for lerna

boneskull committed:
update node.js workflow for lerna

jlipps commented:
Let me know once you get to a place where you're reasonably sure your changes aren't causing test failures and I'll help investigate any. There's some pretty wild stuff that happens in test setup in some places that may be making various assumptions and need rethinking.

jlipps reviewed (commented):
things generally look great here. one thought is that the docs can be moved to the root outside packages (potentially with some of their own tooling). it's not a package and not related to anything else. more like a set of dev scripts. or if we ultimately make the docs a sort of template app that allows driver/plugin authors to publish their own docs, it would be its own package. either way just a note to flag a future improvement of moving docs generation etc out of the main appium package.

boneskull committed:
fix: do not assume git root and package.json have the same parent dir

boneskull committed:
chore: add babel config for debugging (wip)

boneskull committed:
chore: add appium-fake-driver to monorepo; fix appium tests

- all tests changed were needed because of path issues.  I was unable to get the tests working in their entirety _without_ adding `@appium/fake-driver` to the monorepo--it would have required a separate `npm install` step in `appium-fake-driver` due to the lack of its `devDependencies` and `prepare` script--so I figured I might as well pull it in now.
- deduped `package.json` scripts somewhat
- updated ESLint configuration to include `@appium/fake-driver`'s rule override
- sorted the root `package.json` because fussy
- added `wd` as a dev dep because `@appium/fake-driver` needs it

boneskull commented on lerna.json:
```diff
@@ -0,0 +1,26 @@
+  "packages": [
+    "packages/*"
+  ],
+  "version": "2.0.0-beta.1",
```
this only applies if we are not going to version the subpackages separately.  I basically just copied it from what Christian did.  unless there's a real need (shrinkwrap, perhaps?), keeping the versions in sync will probably be more trouble than it's worth (lerna advises against this).

the publishing flow is still TBD anyway--I am hoping to defer the decision for now

boneskull commented on scripts/release-branch.sh:
```diff
@@ -1,13 +0,0 @@
-git fetch
```
okay, I can revert them, but the scripts seem to point to directories that don't exist, so I figured it was unused

boneskull committed:
fix: doc generation

I think?

boneskull commented:
hey, build's green!

boneskull committed:
chore: add @appium/gulp-plugins to monorepo

boneskull committed:
chore: add @appium/doctor to the monorepo

boneskull committed:
chore: fixup a couple READMEs

- will need to deal w/ the `appium` pkg readme
- fix pkg name in `@appium/gulp-plugins`

boneskull committed:
chore: add some convenience scripts for testing individual packages

boneskull committed:
fix: fix the lint:fix script; run parallel

boneskull committed:
chore: add @appium/eslint-config-appium to the monorepo

this pkg is unique in that it does not use `gulp`, so for now we tell lerna to always ignore it when running `lerna exec` across packages.  it has no tests, but it's marginally useful to lint the `index.js`, so we'll do that (see `lint` script in root `package.json`).

the canonical way to handle plugin deps in a sharable config is to use `peerDependencies`, so it now uses `peerDependencies` instead of `dependencies`.  these peer deps are installed by the dependencies of `@appium-gulp-plugins`.  see [docs](https://eslint.org/docs/developer-guide/shareable-configs#publishing-a-shareable-config)

our root `.eslintrc` depends upon `@appium/eslint-config-appium` (which is not an unusual situation).  to pull this off in a monorepo, a cheat is to `file:` schema in the dependency specifier of the root `package.json`.

also tweaked `lerna.json` to _always_ run `bootstrap` with the `--no-ci` flag.  lerna will attempt to use `npm ci` if `CI` env var is present, but for that to work, we must have `package-lock.json` under version control.

boneskull committed:
fix: module/file resolution issues

jlipps commented:
Just thinking of some other tasks for this project in case you haven't got them listed:

- [ ] deprecate any old NPM packages that have been moved to the `@appium` scope
- [ ] add a 'Repo moved' notice to the top of the readme of all the ported repos
- [ ] move the old repos to the appium-boneyard org to help keep our org clean

boneskull committed:
chore: add appium-support to repo

- add `npm start` script to run appium server
- add `npm run doctor` convenience script to run doctor
- renamed `clean` script to `reinstall`
- added new `clean` script to just clean
- updated the "appium from source" doc

boneskull committed:
chore: replace call to lerna in workflow with npm script

boneskull committed:
chore: disable "exit: true" in unit test config; force color

`exit: true` is only necessary in mocha if your code doesn't clean up after itself.  this is not needed for the unit tests--haven't tested with e2e tests yet

boneskull committed:
fix: e2e-tests

- re-add missing fake-plugin dev dep
- export a `PROJECT_ROOT` from the `helpers` module; use this where needed (DRY)

boneskull committed:
chore: enable debugging of tests outside of gulp context

- tests cannot be debugged very easily due to the babel dependencies _and_ the root babel config living in two separate places.  this change adds them to the monorepo root.  the drawback here is that these modules are duplicated, because they live as dependencies in gulp-plugins as well.
- add a bunch of convenience scripts for e2e tests
- normalize lerna args

boneskull commented:
I think we'll want to pull `fake-driver` in here too, then after that I'll merge the history for the base driver.

boneskull commented:
wait, guess not. it's somewhere else

boneskull committed:
chore: prepare to merge history w/ monorepo

boneskull committed:
Merge branch '2.0-monorepo' of ../appium-base-driver into appium-base-driver

* '2.0-monorepo' of ../appium-base-driver: (762 commits)
  chore: prepare to merge history w/ monorepo
  chore: bump 8.0 beta to 8.0.0-beta.6
  refactor!: generalize concepts of server updater and extra method map basedriver shouldn't really know about plugins or drivers. it should just expose the ability to update the server or add methods
  chore: bump 2.0 to beta.5
  fix!: remove image element proxy check, and allow proxyRouteIsAvoided to inspect the body
  chore: bump to 8.0.0-beta.4
  fix: update plugin server adjustment to use new style
  chore: bump to 8.0.0-beta.3
  feat: allow retrieval of basedriver version from any inheriting driver
  chore: bump to 8.0.0-beta.2 since beta.1 is already published
  chore!: remove image element code since it is now handled by @appium/images-plugin
  fix: Avoid unhandled error on driver exit (#447) (#448)
  7.4.1
  fix: Avoid unhandled error on driver exit (#447)
  feat: make vendor prefix mandatory (#440)
  7.4.0
  feat: make server connection timeout configurable
  chore: bump major version
  chore: merge master into 2.0 (#441)
  chore: merge master to 2.0 branch (#438)
  ...

boneskull committed:
chore: cleanup base-driver

- remove dupe files
- move dev deps into root
- move `.azure-pipelines.yml` to root for reasons
- tweak gulp `lint` task to avoid linting `.min.js` files (vendored jQuery)
- update `.eslintignore` to do the same

boneskull committed:
chore: replace precommit hook with eslint --fix

- by way of `lint-staged`

boneskull committed:
fix: use random test port

- update references to `@appium/base-driver`

boneskull committed:
chore: consolidate more eslint config

boneskull committed:
fix(appium): use proper base driver package & random ports

boneskull committed:
chore: add @babel/eslint-parser to root devDeps to enable in-IDE linting

boneskull committed:
fix(base-driver): use random ports in tests

boneskull committed:
chore: add vscode settings to .gitignore

boneskull committed:
fix(fake-driver): use appium-base-driver@1.x for now

boneskull committed:
chore: add @appium/test-support to monorepo & use Bluebird w/ Sinon

sinon supports promises natively, and we are using bluebird here.  promise resolution seems to go wonky when using sinon's stub `.returns(B.reject(err))`, so now we tell sinon to use _bluebird's_ promises, which makes `.rejects(err)` do the same thing except better.

boneskull committed:
fix(doctor): use @appium/test-support

boneskull committed:
fix(base-driver): more random ports

boneskull committed:
chore: remove unused e2e test scripts; add test-support test script

boneskull committed:
fix(docs): re-add appium-base-driver as devdep for docgen

boneskull committed:
chore: revert removal of scripts folder

boneskull commented:
# UPDATE

## packages now in monorepo

- `appium`
- `@appium/base-driver`
- `@appium/doctor`
- `@appium/eslint-config-appium`
- `@appium/fake-driver`
- `@appium/gulp-plugins`
- `@appium/support`
- `@appium/test-support`

All package names have changed _except_ `appium`.

## Notes

- The entire Git history from `appium-base-driver` is part of this PR.  Before merging, I [moved _all_ files in this project into a subdir, `packages/base-driver`](https://github.com/appium/appium/pull/15343/commits/a55d3136341a05b602939f08d90d6671ea540324). 

- Some devDeps are duplicated between the root monorepo (where _all_ dev deps should live) and `@appium/gulp-plugins`, which wants to provide its own as _dependencies_.  Moving these dev deps into the root enables IDE integration (debugging, eslint, mocha, etc), whereas an IDE will have a difficult time finding the dev deps in `packages/gulp-plugins/node_modules`.

- Lerna's "hoist" mode does not seem to improve installation time.

- The `prepare` scripts (transpilation) to be run for each package (except the eslint config) mean npm workspaces cannot be used at this time (or likely ever).  npm workspaces work differently than Lerna's `bootstrap`--in Lerna, the root packages are installed, _then_ the leaf packages are installed.  npm installs leaves _first_ (like a depth-first search), which means the dev deps wouldn't be in place anyway to run `prepare`.  On the one hand, forcing a build upon install isn't really ideal, but neither is npm's behavior. `¯\_(ツ)_/¯` 

## Future Work (Not part of this PR)

- Figure out what to do with the `README`s
- Design & document new publishing workflow
- Refactor doc gen
- Consolidate or remove Gulp entirely; dedupe dev deps
- Future of `@appium/test-support`?
- Build & test optimizations

jlipps commented:
Another task for later:
* update all packages which depend on the moved packages to use the new name (and push a deprecation notice to the old names?). This should be the case for all the moved packages except for base driver (since the basedriver for Appium 1.x has not been moved, and can still be developed/published as appium-base-driver, while we continue appium 2.0 basedriver work with `@appium/base-driver`

boneskull committed:
chore: remove unused azure pipelines config

boneskull commented on .eslintrc:
```diff
@@ -1,3 +1,14 @@
+      "files": "packages/*/test/**/*",
+      "rules": { "func-names": "off" }
+    }
+  ]
```
@jlipps 

this file consolidates all `.eslintrc` files across the packages.

however, the `lint` gulp task also does whatever it does.  I'm a little concerned `eslint --fix` won't have the same result as `gulp lint --fix`.  it's _probably_ not a problem, but we should keep an eye open for linting churn.

boneskull commented on babel.config.json:
```diff
@@ -0,0 +1,39 @@
+{
```
@jlipps

`babel.config.json` (or `babel.config.js`) is the "root" babel config as of Babel v7.  a `.babelrc` file in any package will merge or override (not clear...) the two config files.  

this is roughly the same as the `.babelrc` file in `gulp-plugins` -- but trying to output some source maps here for debugging & stack traces.  I'm _pretty sure_ this file is not yet used anywhere _except_ when debugging in an IDE.

boneskull committed:
chore: remove invalid comment

boneskull committed:
chore(base-driver): use random port in websockets e2e test

boneskull committed:
chore(fake-driver): add convenience script for e2e tests

boneskull committed:
chore(base-driver): use random port in express server tests

mykola-mokhnach commented on babel.config.json:
```diff
@@ -0,0 +1,39 @@
+      "@babel/preset-env",
+      {
+        "targets": {
+          "node": "10"
```
it makes sense to bump it to 12 now. Probably, in another PR

boneskull replied:
see https://github.com/appium/appium/issues/15401

mykola-mokhnach commented on docs/en/contributing-to-appium/appium-from-source.md:
```diff
@@ -11,10 +11,11 @@ request; for more information on how to run tests, keep reading!
 * N - [https://github.com/tj/n](https://github.com/tj/n)
 
-Your Node.js installation will include the [NPM](https://www.npmjs.com/) package manager, which Appium
+Your Node.js installation will include the [npm](https://www.npmjs.com/) package manager, which Appium
```
version 6+

mykola-mokhnach commented on docs/en/contributing-to-appium/appium-from-source.md:
```diff
@@ -174,4 +156,4 @@ the commit from occurring.
 Once code is committed and a [pull request](https://help.github.com/articles/about-pull-requests/)
 is made to the correct Appium respository on [GitHub](https://github.com/), Appium build system
-will run all of the functional tests.
+will run all of the functional tests. 
```
nit: trailing spaces

boneskull replied:
we should adopt prettier, or something 😄

mykola-mokhnach replied:
I  have simply enabled trimming in my VSCode

boneskull replied:
trailing spaces are significant in github-flavored markdown, so that might be bad.  even with an `.editorconfig` file -- which most editors/IDEs support -- you'd have to skip trimming markdown.  prettier is smart enough to know how to handle this.

mykola-mokhnach commented on package.json:
```diff
@@ -1,112 +1,98 @@
+    "serve-static": "^1.13.1",
     "sinon": "^9.0.0",
     "validate.js": "^0.13.0",
+    "wd": "^1.14.0",
```
note: this is something we don't want to depend anymore :)

boneskull replied:
yeah, need to update fake-driver.  that will be a separate PR

mykola-mokhnach commented on package.json:
```diff
@@ -1,112 +1,98 @@
 {
```
does dependabot handle monorepo dependencies updates properly?

boneskull replied:
¯\_(ツ)_/¯ I'll research this

boneskull replied:
it looks like it doesn't handle this very well at present time.  

renovate does, however, and I think that provides something similar. I'll investigate in a subsequent PR.

boneskull replied:
see https://github.com/appium/appium/issues/15400

mykola-mokhnach commented on packages/appium/gulpfile.js:
```diff
@@ -47,8 +46,7 @@ boilerplate({
     files: ['${testDir}/**/*-specs.js']
   },
-  testTimeout: 160000,
-  preCommitTasks: ['eslint', 'once'],
```
so we got rid of all precommit tasks or we still keep the eslint one?

boneskull replied:
the only precommit task is now eslint, and it's defined in `package.json` only

mykola-mokhnach commented on packages/appium/lib/config.js:
```diff
@@ -44,11 +44,23 @@ async function updateBuildInfo (useGithubApiFallback = false) {
+ * Finds the Git metadata dir (see `GIT_META_ROOT`)
+ *
+ * This is needed because Appium cannot assume `package.json` and `.git` are in the same
+ * directory.  Monorepos, see?
```
🙈

mykola-mokhnach commented on packages/appium/lib/config.js:
```diff
@@ -44,11 +44,23 @@ async function updateBuildInfo (useGithubApiFallback = false) {
+ *
+ * This is needed because Appium cannot assume `package.json` and `.git` are in the same
+ * directory.  Monorepos, see?
+ * @returns {string|void} Path to dir or `undefined` if not found
```
we usually put it as `?string`

boneskull replied:
we should lint the docstrings 😄

boneskull replied:
`?string` means "`string` or `null`", and this function never returns `null`.

mykola-mokhnach replied:
this function might return undefined though

jlipps replied:
isn't that what `void` means?

mykola-mokhnach commented on packages/appium/lib/config.js:
```diff
@@ -44,11 +44,23 @@ async function updateBuildInfo (useGithubApiFallback = false) {
+ * @returns {string|void} Path to dir or `undefined` if not found
+ */
+async function findGitRoot () {
+  return await findUp(GIT_META_ROOT, {cwd: rootDir, type: 'directory'});
```
cannot we use the existing glob package from appium-support for this purpose instead of adding a new dependency?

boneskull replied:
I mean, we could, but we'd end up writing our own `find-up`.

boneskull replied:
afaict there's no wildcard that tells glob/minimatch to navigate to an arbitrary parent directory from cwd.  you'd need to start from the fs root and work forwards, which would be...slow.  so you'd need to search for `.git` using glob, and if it's not there, navigate to `..`, then search again (but skip the directory we just came from), and do so in a loop.  I imagine this is basically what `find-up` is doing

mykola-mokhnach replied:
np, I'm just usually being a bit paranoidal about adding new dependecies because some of them might exponentially increase the size of the linked modules tree %)

mykola-mokhnach commented on packages/appium/test/cli-e2e-specs.js:
```diff
@@ -4,10 +4,14 @@ import chai from 'chai';
 chai.use(chaiAsPromised);
-const cwd = path.resolve(__dirname, '..', '..');
+
+// cannot use `require.resolve()` here (w/o acrobatics) due to the ESM context.
```
acrobatics 😄

boneskull replied:
not sure if it's even possible with babel.  babel lets you do things you couldn't otherwise, but I don't think it supports all of Node.js' native ESM handling

mykola-mokhnach commented on packages/appium/test/helpers.js:
```diff
@@ -0,0 +1,35 @@
+import findUp from 'find-up';
+import getPort from 'get-port';
+
+const TEST_HOST = 'localhost';
```
I would rather prefer 127.0.0.1
Some hosts may resolve to ipv6 address by default, which we don't want

boneskull replied:
actually, I think I'm going to leave it, since `localhost` is hardcoded throughout the tests and it will be a pretty sizable diff.

boneskull replied:
see https://github.com/appium/appium/issues/15399

mykola-mokhnach commented on packages/appium/test/helpers.js:
```diff
@@ -0,0 +1,35 @@
+let TEST_PORT;
+/**
+ * Returns a free port; one per process
+ * @returns {Promise<number>} a free port
```
we skip `Promise` and only put the actual value type, since an IDE knows anyway that the result of an `async` function is always a promise

boneskull replied:
that's true that the IDE knows, but if we're being explicit about `return await` it would be consistent to be explicit about this, yes?

mykola-mokhnach replied:
I never use Promise notation anywhere for async returns. Yolo :)

jlipps replied:
we should just port everything to typescript and be done with it :trollface:

boneskull replied:
@jlipps imo it's not needed.  properly-written docstrings are easily understood by typescript, and you can just use it as a type linter (and you get that info in the IDE!).   I wrote [report-toolkit](https://github.com/ibm/report-toolkit) this way.

mykola-mokhnach commented on packages/appium/test/plugin-e2e-specs.js:
```diff
@@ -8,26 +8,31 @@ import { remote as wdio } from 'webdriverio';
+
+const FAKE_PLUGIN_DIR = path.join(PROJECT_ROOT, 'node_modules', '@appium', 'fake-plugin');
+const FAKE_DRIVER_DIR = path.join(PROJECT_ROOT, 'packages', 'fake-driver');
+let TEST_SERVER;
```
I would prefer to define variables closer to the scope where they are used (`describe` in such case)

boneskull replied:
ok

boneskull commented:
The CLA check is "pending" because apparently some _other_ commit authors have not yet signed it.  Or at least, that's my best guess.  The bot is dumb.

boneskull committed:
chore(appium): tweak declarations and var names in plugin-e2e-specs test

boneskull committed:
docs: whitespace

boneskull committed:
docs: fix required npm version number

christian-bromann reviewed (commented):
Some comments:

- `packages/base-driver/lib/basedriver/...` vs. `packages/basedriver/lib/basedriver/desired-caps.js `?
- is it worth maybe moving appium-express server package into its own sub package or will it be completely absorbed by the basedriver? Maybe it makes sense anyway, my feeling is that if we can keep sub packages reasonable small it simplifies contributing to it - it also helps in cases e.g. where we want to migrate technologies like move to TypeScript where it allows us to migrate package by package
